master/buildbot/util/service.py

```diff
@@ -462,23 +465,17 @@ def reconfigServiceWithBuildbotConfig(self, new_config):
         # calculate new childs, by name, and removed childs
         removed_names, added_names = util.diffSets(old_set, new_set)
 
-        # find any childs for which the fully qualified class name has
-        # changed, and treat those as an add and remove
-        # While we're at it find any service that don't know how to reconfig,
-        # and, if they have changed, add them to both removed and added, so that we
+        # find any children for which the old instance is not
+        # able to do a reconfig with the new sibling
+        # and add them to both removed and added, so that we
         # run the new version
         for n in old_set & new_set:
             old = old_by_name[n]
             new = new_by_name[n]
-            # detect changed class name
-            if reflect.qual(old.__class__) != reflect.qual(new.__class__):
+            # check if we are able to reconfig service
+            if not old.canReconfigWithSibling(new):
                 removed_names.add(n)
                 added_names.add(n)
-            # compare using ComparableMixin if they don't support reconfig
-            elif not hasattr(old, 'reconfigServiceWithBuildbotConfig'):
```

Comment:
Could we move this change to a separate PR?

It would be great to add this to the description of the commit so that any future readers are less confused:

```
Child services of BuildbotServiceManager can't have reconfigServiceWithBuildbotConfig and even if they do it does not do anything because it may only be invoked by the parent service. In this case parent service is BuildbotServiceManager and it does not invoke reconfigServiceWithBuildbotConfig.

In the end this code made reconfiguration less efficient because services were added to removed_names and added_names unnecessarily.
```

Reply:
I'm not sure I understand that comment. This line was related to (I believe) old classes that were not subclassing `BuildbotService` and so didn't have the `reconfigServiceWithBuildbotConfig` method and couldn't be reconfigured.

But the assumption from removing this check is all services can be reconfigured, and so we don't need to check the existence of this method anymore.

Reply:
OK, you're right, I missed that `BuildbotService` derives from `ReconfigurableServiceMixin`. So perhaps the old code was using `reconfigServiceWithBuildbotConfig` as a proxy for existence of `reconfigServiceWithSibling` which is what `BuildbotService` is actually calling later.

Then maybe this sentence within the description of the commit would make sense?

"""
All services handled by BuildbotServiceManager are now expected to derive from BuildbotService and thus will have both reconfigServiceWithBuildbotConfig and reconfigServiceWithSibling. Therefore the check is no longer needed.
"""

I just want to have some more information of what's happened because if this commit breaks stuff, the explanation for the breakage could better be in its description and not in some PR comments that are harder to find.

Reply:
Yes, I think that comment would make sense, I'll add that to the commit message.